074-k8s-storage/proposal.md

```diff
@@ -0,0 +1,167 @@
+# Terms
+- **cache object** a BLOB and relevant metadata that Concourse needs to persist. These could be Resource Caches, Task Caches or Build Caches.
+- **worker** Concourse executes steps on a **worker** and implements some **worker** interface. Concourse is agnostic of the runtime representation of the worker (eg. K8s pod, node or cluster).
+
+# Summary
+
+After spiking on a few solutions for storage on Kubernetes our recommendation is to use an image registry to store **cache objects** for steps.
+
+# Motivation
+
+As we started thinking about the Kubernetes runtime we realized that we need to think about what our storage solution would be before proceeding with any other part of the implementation. Storage has a huge effect on how Concourse interacts with the runtime (Kubernetes). Storage also had a lot of unknowns, we didn't know what the storage landscape on Kubernetes looked like and what options were available to us. Storage also has a huge impact on the performance of the cluster, in regards to storage and initialization of steps.
+
+## Requirements
+An ideal storage solution can do the following :
+
+- image fetching from the CRI k8s is using
+- transfer **cache objects** between steps (whatever represents a step, most likely a pod)
+- cache for resources and tasks
+- stream **cache objects** across worker runtimes (k8s worker sends artifact to garden worker)
+
+## Criteria
+- security
+- performance, aka initialization time (time spent running workloads on a single k8s worker, as well as across workers)
+- resource usage to run this storage solution
+
+# Proposal
+
+**TL;DR**: We recommend going with the image registry option because it satisfies all the requirements and gives us a bunch of options to improve performance when compared to the blobstore option. It also provides a very flexible solution that works across multiple runtime workers. [See Details](#image-registry-to-store-artifacts)
```

Comment:
Trade-study - push/pull from registry using
1. init & side car containers
1. CSI with independent volume/input (eg. image populator)
1. Custom image per combination of step for pulling and TODO for pushing